lib/events/azsessions/azsessions.go

```diff
@@ -0,0 +1,573 @@
+	inprogressContainerParam = "inprogress_container"
+)
+
+// defaultSessionContainerName and defaultInprogressContainerName are the
+// default container names for finalized session recordings and for in-progress
+// data.
+const (
+	defaultSessionContainerName    = "session"
+	defaultInprogressContainerName = "inprogress"
+)
+
+// sessionName returns the name of the blob that contains the recording for a
+// given session.
+func sessionName(sid session.ID) string {
+	return sid.String()
+}
+
+// uploadMarkerPrefix is the prefix of the names of the upload marker blobs.
+// Listing blobs with this prefix will return an empty blob for each upload.
+const uploadMarkerPrefix = "upload/"
+
+// uploadMarkerName returns the blob name for the marker for a given upload.
+func uploadMarkerName(upload events.StreamUpload) string {
+	return fmt.Sprintf("%v%v/%v", uploadMarkerPrefix, upload.SessionID, upload.ID)
+}
+
+// partPrefix returns the prefix for the upload part blobs for a given upload.
+// Listing blobs with this prefix will return all the parts that currently make
+// up the upload.
+func partPrefix(upload events.StreamUpload) string {
+	return fmt.Sprintf("part/%v/%v/", upload.SessionID, upload.ID)
+}
+
+// partName returns the name of the blob for a specific part in an upload.
+func partName(upload events.StreamUpload, partNumber int64) string {
+	return fmt.Sprintf("%v%v", partPrefix(upload), partNumber)
+}
+
+// field names used for logging
+const (
+	fieldSessionID  = "session_id"
+	fieldUploadID   = "upload_id"
+	fieldPartNumber = "part"
+	fieldPartCount  = "parts"
+)
+
+// Config is a struct of parameters to define the behavior of Handler.
+type Config struct {
+	// ServiceURL is the URL for the storage account to use.
+	ServiceURL url.URL
+
+	// SessionContainerName is the name of the container that stores finalized
+	// session recordings. Defaults to [defaultSessionContainerName].
+	SessionContainerName string
+
+	// InprogressContainerName is the name of the container that stores
+	// in-progress data that's yet to be finalized in a recording. Defaults to
+	// [defaultInprogressContainerName].
+	InprogressContainerName string
+
+	// ClientID, when set, defines the managed identity's client ID to use for
+	// authentication.
+	ClientID string
+
+	// Log is the logger to use. If unset, it will default to the global logger
+	// with a component of "azblob".
+	Log logrus.FieldLogger
+}
+
+// SetFromURL sets values in Config based on the passed in URL: the fragment of
+// the URL is parsed as if it was made out of query parameters, which define
+// options for ourselves, and then the remainder of the URL is set as the
+// service URL.
+func (c *Config) SetFromURL(u *url.URL) error {
+	if u == nil {
+		return nil
+	}
+
+	c.ServiceURL = *u
+
+	switch c.ServiceURL.Scheme {
+	case teleport.SchemeAZBlob:
+		c.ServiceURL.Scheme = "https"
+	case teleport.SchemeAZBlobHTTP:
+		c.ServiceURL.Scheme = "http"
+	}
```

Comment:
In other cases we just leave the scheme as is, so you can spin up tests with the https URL you copied off of the Azure portal - Teleport itself will only route you here if the scheme is `azblob` or `azblob-http` anyway. I'm not sure if there's any value in checking the scheme here rather than letting it fail later.

Reply:
I went back to this, now we're enforcing that the scheme is `azblob`, the undocumented `azblob-http`, `https` or `http` (the last two only usable in tests or custom code) because the blob storage SDK wouldn't know what to do with a non-http URL anyway.